File: src/tools/interop/idt/utils/log.py
#
#    Copyright (c) 2023 Project CHIP Authors
#    All rights reserved.
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
#

import logging
from typing import TextIO

from termcolor import colored

import config

_CONFIG_LEVEL = config.log_level

_FORMAT_PRE_FSTRING = "%(asctime)s %(levelname)s {%(module)s} [%(funcName)s] "
_FORMAT_PRE = colored(_FORMAT_PRE_FSTRING, "blue") if config.enable_color else _FORMAT_PRE_FSTRING
_FORMAT_POST = "%(message)s"
_FORMAT_NO_COLOR = _FORMAT_PRE_FSTRING+_FORMAT_POST

FORMATS = {
    logging.DEBUG: _FORMAT_PRE + colored(_FORMAT_POST, "blue"),
    logging.INFO: _FORMAT_PRE + colored(_FORMAT_POST, "green"),
    logging.WARNING: _FORMAT_PRE + colored(_FORMAT_POST, "yellow"),
    logging.ERROR: _FORMAT_PRE + colored(_FORMAT_POST, "red", attrs=["bold"]),
    logging.CRITICAL: _FORMAT_PRE + colored(_FORMAT_POST, "red",  "on_yellow", attrs=["bold"]),
}


class LoggingFormatter(logging.Formatter):

    def format(self, record):
        log_fmt = FORMATS.get(record.levelno) if config.enable_color else _FORMAT_NO_COLOR
        formatter = logging.Formatter(log_fmt)
        return formatter.format(record)


def get_logger(logger_name) -> logging.Logger:
    logger = logging.getLogger(logger_name)
    logger.setLevel(_CONFIG_LEVEL)
    ch = logging.StreamHandler()
    ch.setLevel(_CONFIG_LEVEL)
    ch.setFormatter(LoggingFormatter())
    logger.addHandler(ch)
    logger.propagate = False
    return logger


def border_print(to_print: str, important: bool = False) -> None:
    len_borders = len(to_print)
    border = f"\n{'_' * len_borders}\n"
    i_border = f"\n{'!' * len_borders}\n" if important else ""
    to_print = f"{border}{i_border}{to_print}{i_border}{border}"
    if config.enable_color:
        to_print = colored(to_print, "magenta")
    print(to_print)


def print_and_write(to_print: str, file: TextIO) -> None:
    if config.enable_color:
        print(colored(to_print, "green"))
    else:
        print(to_print)
    file.write(to_print)


def add_border(to_print: str) -> str:
    return '\n' + '*' * len(to_print) + '\n' + to_print

File: examples/common/tracing/decoder/logging/Log.cpp
/*
 *   Copyright (c) 2022 Project CHIP Authors
 *   All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

#include "Log.h"

#include "ToCertificateString.h"

#include <lib/support/BytesToHex.h>
#include <lib/support/SafeInt.h>

namespace chip {
namespace trace {
namespace logging {

namespace {

constexpr uint8_t kSpacePerIndent = 4;
uint8_t gIndentLevel              = 0;

void ENFORCE_FORMAT(1, 2) LogFormatted(const char * format, ...)
{
    char buffer[CHIP_CONFIG_LOG_MESSAGE_MAX_SIZE + 1] = {};

    int indentation = gIndentLevel * kSpacePerIndent;
    snprintf(buffer, sizeof(buffer), "%*s", indentation, "");

    va_list args;
    va_start(args, format);
    vsnprintf(&buffer[indentation], sizeof(buffer) - static_cast<size_t>(indentation), format, args);
    va_end(args);

    ChipLogDetail(DataManagement, "%s", buffer);
}

const char * ToHexString(chip::ByteSpan source, MutableCharSpan destination)
{
    memset(destination.data(), '\0', destination.size());

    if (source.size() == 0)
    {
        return destination.data();
    }

    CHIP_ERROR err =
        Encoding::BytesToHex(source.data(), source.size(), destination.data(), destination.size(), Encoding::HexFlags::kUppercase);
    if (CHIP_NO_ERROR != err)
    {
        LogErrorOnFailure(err);
        return destination.data();
    }

    return destination.data();
}

} // namespace

void Log(const char * name, ByteSpan & data)
{
    if (data.empty())
    {
        return;
    }

    // If the size of the data is larger than half of the maximum size for a log message (minus 1 for null-termination),
    // reduce the size of the data to fit within this limit.
    // The limit is half the size of the message because we will be logging a hex representation of that data, at 2 chars per byte.
    if (data.size() > (CHIP_CONFIG_LOG_MESSAGE_MAX_SIZE / 2) - 1)
    {
        data.reduce_size((CHIP_CONFIG_LOG_MESSAGE_MAX_SIZE / 2) - 1);
    }

    char buffer[CHIP_CONFIG_LOG_MESSAGE_MAX_SIZE] = { 0 };
    chip::MutableCharSpan destination(buffer);

    // Check if the size of the data can be cast to uint16_t.
    // If yes, log the name and size of the data followed by its hexadecimal string representation.
    // If no, log the name and a marker indicating the size exceeds UINT16_MAX, followed by the data's hexadecimal string.
    CanCastTo<uint16_t>(data.size())
        ? LogFormatted("%s (%u) = %s", name, static_cast<uint16_t>(data.size()), ToHexString(data, destination))
        : LogFormatted("%s (>UINT16_MAX) = %s", name, ToHexString(data, destination));
}

void Log(const char * name, CharSpan & data)
{
    if (data.empty())
    {
        return;
    }

    CanCastTo<uint16_t>(data.size())
        ? LogFormatted("%s (%u) = %.*s", name, static_cast<uint16_t>(data.size()), static_cast<int>(data.size()), data.data())
        : LogFormatted("%s (>UINT16_MAX) = %.*s", name, static_cast<int>(data.size()), data.data());
}

void Log(const char * name, uint16_t value)
{
    LogFormatted("%s = %u", name, value);
}

void Log(const char * name, uint32_t value)
{
    LogFormatted("%s = %u", name, value);
}

void Log(const char * name, const char * value)
{
    LogFormatted("%s = %s", name, value);
}

void Log(const char * name)
{
    LogFormatted("%s", name);
}

void LogAsHex(const char * name, uint8_t value)
{
    LogFormatted("%s = 0x%02x", name, value);
}

void LogAsHex(const char * name, uint16_t value)
{
    LogFormatted("%s = 0x%04x", name, value);
}

void LogAsHex(const char * name, uint64_t value)
{
    LogFormatted("%s = " ChipLogFormatX64, name, ChipLogValueX64(value));
}

void LogCertificate(const char * name, const ByteSpan & data)
{
    char buffer[CHIP_CONFIG_LOG_MESSAGE_MAX_SIZE] = {};
    chip::MutableCharSpan destination(buffer);

    CanCastTo<uint16_t>(data.size()) ? LogFormatted("%s (%u) =", name, static_cast<uint16_t>(data.size()))
                                     : LogFormatted("%s (>UINT16_MAX) =", name);
    LogFormatted("{\n%s", ToCertificateString(data, destination));
    LogFormatted("}");
    LogFormatted(" ");
}

void LogCertificateRequest(const char * name, const ByteSpan & data)
{
    char buffer[CHIP_CONFIG_LOG_MESSAGE_MAX_SIZE] = {};
    chip::MutableCharSpan destination(buffer);

    CanCastTo<uint16_t>(data.size()) ? LogFormatted("%s (%u) =", name, static_cast<uint16_t>(data.size()))
                                     : LogFormatted("%s (>UINT16_MAX) =", name);
    LogFormatted("{\n%s", ToCertificateRequestString(data, destination));
    LogFormatted("}");
    LogFormatted(" ");
}

void IncreaseLogIndent()
{
    LogFormatted("{");
    gIndentLevel++;
}

void DecreaseLogIndent()
{
    gIndentLevel--;
    LogFormatted("}");

    if (gIndentLevel == 0)
    {
        LogFormatted(" ");
    }
}

ScopedLogIndent::ScopedLogIndent(const char * name)
{
    LogFormatted("%s =", name);
    IncreaseLogIndent();
}

ScopedLogIndent::~ScopedLogIndent()
{
    DecreaseLogIndent();
}

ScopedLogIndentWithFlags::ScopedLogIndentWithFlags(const char * name, uint8_t flags)
{
    LogFormatted("%s (0x%02x) =", name, flags);
    IncreaseLogIndent();
}

ScopedLogIndentWithFlags::~ScopedLogIndentWithFlags()
{
    DecreaseLogIndent();
}

ScopedLogIndentWithSize::ScopedLogIndentWithSize(const char * name, size_t size)
{
    CanCastTo<uint16_t>(size) ? LogFormatted("%s (%u bytes) =", name, static_cast<uint16_t>(size))
                              : LogFormatted("%s (>UINT16_MAX) =", name);
    IncreaseLogIndent();
}

ScopedLogIndentWithSize::~ScopedLogIndentWithSize()
{
    DecreaseLogIndent();
}

} // namespace logging
} // namespace trace
} // namespace chip
